src/FSharp.Formatting.Literate/ParsePynb.fs

```diff
@@ -0,0 +1,162 @@
+namespace FSharp.Formatting.Literate
+
+open System.IO
+open System.Text.Json
+open FSharp.Formatting.Templating
+open FSharp.Formatting.PynbModel
+
+module internal ParsePynb =
+
+    type ParsedCell =
+        | Code of
+            {| lang: string
+               source: string
+               outputs: string[] option |}
+        | Markdown of source: string
+
+        member this.ToMarkdown() =
+            match this with
+            | Markdown source -> source
+            | Code code ->
+                let codeBlock = sprintf $"```{code.lang}\n{addLineEnd code.source}```"
+
+                match code.outputs with
+                | None -> codeBlock
+                | Some outputs ->
+                    let outputsString = outputs |> String.concat "\n"
+                    sprintf $"{codeBlock}\n{outputsString}"
+
+    module Output =
+        let (|TextHtml|_|) (x: JsonElement) =
+            match x.TryGetProperty("text/html") with
+            | true, html ->
+                let html = html.EnumerateArray() |> Seq.map (fun x -> x.GetString()) |> String.concat "\n"
+                Some("""<p>""" + html + """</p>""")
+            | _ -> None
+
+        let (|TextPlain|_|) (x: JsonElement) =
+            match x.TryGetProperty("text/plain") with
+            | true, text ->
+                let text = text.EnumerateArray() |> Seq.map (fun x -> x.GetString()) |> String.concat ""
+
+                Some(
+                    """<table class="pre"><tbody><tr><td><pre><code>"""
+                    + text
+                    + """</code></pre></td></tr></tbody></table>"""
+                )
+            | _ -> None
+
+        let (|DisplayData|_|) (x: JsonElement) =
+            if x.GetProperty("output_type").GetString() = "display_data" then
```

Comment:
Would it not make sense to be a little bit more defensive in these active patterns?
Use `x.TryGetProperty` to avoid a null pointer if `"output_type"` is not present?